db_design.md

```diff
@@ -0,0 +1,170 @@
+# Credit Tracking Service Design Document
+
+## Overview
+
+The Credit Tracking Service manages developer credits for API access through a gRPC interface. It handles credit deduction, automatic top-ups via blockchain token burns, and maintains transaction state for unconfirmed burns.
+
+## Service Flow
+
+### 1. API Access & Credit Deduction
+
+```
+Telemetry API → Credit Tracker (gRPC) → Database
+"Deduct 10 credits for license X accessing asset Y"
+```
+
+- External APIs (telemetry, location, etc.) contact credit tracker via gRPC
+- Credit tracker deducts from active grants using FIFO logic
+- Returns success/failure to calling API
+
+### 2. Insufficient Funds & Auto Top-Up
+
+```
+Insufficient Credits → Initiate DCX Burn → Pending Grant → Continue Operations
+```
+
+- When credits insufficient, service automatically burns DCX from developer's wallet
+- Creates **pending** grant immediately (before blockchain confirmation)
+- Allows continued API usage while transaction confirms (1-2 minutes)
+- Developer temporarily goes negative until burn confirms
+
+### 3. Transaction State Management
+
+- **Pending**: Transaction submitted, credits usable, developer may be negative
+- **Confirmed**: Transaction finalized on-chain, credits fully backed
+- **Failed**: Transaction reverted, **debt must be cleared before further usage**
```

Comment:
This is good.

Clearly there's a little danger here if we have RPC problems.

Reply:
Maybe we add a Failure reason somewhere so we can determine how to take action.
How does these scenarios sound and can you think of any others we would want to handle.

1. If we fail to check the state because network issues => retry
2. If the block fails to get mined => retry (maybe with more gas)
3. If the execution gets reverted => do nothing